subsys/bluetooth/host/l2cap.c

```diff
@@ -2267,6 +2292,95 @@ static void l2cap_chan_send_credits(struct bt_l2cap_le_chan *chan,
 	LOG_DBG("chan %p credits %lu", chan, atomic_get(&chan->rx.credits));
 }
 
+#ifdef CONFIG_BT_L2CAP_SEG_RECV
+static int l2cap_chan_send_credits_pdu(struct bt_conn *conn, uint16_t cid, uint16_t credits)
+{
+	int err;
+	struct net_buf *buf;
+	struct bt_l2cap_le_credits *ev;
+
+	buf = l2cap_create_le_sig_pdu(NULL, BT_L2CAP_LE_CREDITS, get_ident(), sizeof(*ev));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	ev = net_buf_add(buf, sizeof(*ev));
+	*ev = (struct bt_l2cap_le_credits){
+		.cid = sys_cpu_to_le16(cid),
+		.credits = sys_cpu_to_le16(credits),
+	};
+
+	err = bt_l2cap_send(conn, BT_L2CAP_CID_LE_SIG, buf);
+	if (err) {
+		net_buf_unref(buf);
+		return err;
+	}
+
+	return 0;
+}
+
+/**
+ * Combination of @ref atomic_add and @ref u16_add_overflow. Leaves @p
+ * target unchanged if an overflow would occur. Assumes the current
+ * value of @p target is representable by uint16_t.
+ */
+static bool atomic_add_safe_u16(atomic_t *target, uint16_t addition)
+{
+	uint16_t target_old, target_new;
+
+	do {
+		target_old = atomic_get(target);
+		if (u16_add_overflow(target_old, addition, &target_new)) {
+			return true;
+		}
+	} while (!atomic_cas(target, target_old, target_new));
+
+	return false;
+}
+
+int bt_l2cap_chan_give_credits(struct bt_l2cap_chan *chan, uint16_t additional_credits)
+{
+	struct bt_l2cap_le_chan *le_chan = BT_L2CAP_LE_CHAN(chan);
+
+	if (!chan || !chan->ops) {
+		LOG_ERR("%s: Invalid chan object.", __func__);
+		return -EINVAL;
+	}
+
+	if (!chan->ops->seg_recv) {
+		LOG_ERR("%s: Available only with seg_recv.", __func__);
+		return -EINVAL;
+	}
+
+	if (additional_credits == 0) {
+		LOG_ERR("%s: Refusing to give 0.", __func__);
+		return -EINVAL;
+	}
+
+	if (bt_l2cap_chan_get_state(chan) == BT_L2CAP_CONNECTING) {
+		LOG_ERR("%s: Cannot give credits while connecting.", __func__);
+		return -EBUSY;
+	}
+
+	if (atomic_add_safe_u16(&le_chan->rx.credits, additional_credits)) {
+		LOG_ERR("%s: Overflow.", __func__);
+		return -EOVERFLOW;
+	}
+
+	if (bt_l2cap_chan_get_state(chan) == BT_L2CAP_CONNECTED) {
+		int err;
+
+		err = l2cap_chan_send_credits_pdu(chan->conn, le_chan->rx.cid, additional_credits);
+		if (err) {
+			LOG_ERR("%s: PDU failed %d.", __func__, err);
```

Comment:
Does it make sense to log with `LOG_ERR` when we are also returning the error code to the caller? IMO it makes more sense to use `LOG_DBG` in such situations

Reply:
Same applies for all LOG statements in this function